backend/src/database/migrations/V1720446425__segmentsWithParentIdsAndType.sql

```diff
@@ -0,0 +1,56 @@
+ALTER TABLE "segments" ADD COLUMN IF NOT EXISTS "parentId" uuid;
+ALTER TABLE "segments" ADD COLUMN IF NOT EXISTS "grandparentId" uuid;
+ALTER TABLE "segments" ADD COLUMN "type" TEXT GENERATED ALWAYS AS (
+        CASE
+            WHEN "parentSlug" IS NULL AND "grandparentSlug" IS NULL THEN 'projectGroup'
+            WHEN "grandparentSlug" IS NULL AND "parentSlug" IS NOT NULL THEN 'project'
+            ELSE 'subproject' 
+        END
+    ) STORED;
+
+
+BEGIN;
+DO
+$$
+    DECLARE
+        segment segments%ROWTYPE;
+        parent segments%ROWTYPE;
+        grandparent segments%ROWTYPE;
+    BEGIN
+        FOR segment IN SELECT * FROM segments
+            LOOP
+                IF segment."type" = 'project' THEN
+                    SELECT * INTO parent
+                    FROM segments
+                    WHERE "slug" = segment."parentSlug"
+                    AND "type" = 'projectGroup';
+
+                    UPDATE segments
+                    SET "parentId" = parent.id
+                    WHERE id = segment.id;
+
+
+                ELSIF segment."type" = 'subproject' THEN
+                    SELECT * INTO parent
+                    FROM segments
+                    WHERE "slug" = segment."parentSlug"
+                    AND "type" = 'project';
+
+                    SELECT * INTO grandparent
+                    FROM segments
+                    WHERE "slug" = segment."grandparentSlug"
+                    AND "type" = 'projectGroup';
```

Comment:
Have you tried running this on production? At least without the updates, just with something like
```
RAISE NOTICE 'updating parent id to %', parent.id;
```

I'm afraid this kind of filtering won't be enough. Since we don't have a unique index on `slug`, there might be many different segments on the same level with the same slug. And if we don't check for parent slugs too, we might get inconsistencies.

I think if you just run it on production without updates, and if I'm right about inconsistencies, then queries like
```
                    SELECT * INTO parent
                    FROM segments
                    WHERE "slug" = segment."parentSlug"
                    AND "type" = 'project';
```
will fail because they'll try to select multiple rows into one variable.

Reply:
we have these indexes 
```
CREATE UNIQUE INDEX "segments_slug_pSlug_gpSlugNull_tenantId_idx" ON "segments" ("slug", "parentSlug", "tenantId") WHERE "grandparentSlug" IS NULL;
CREATE UNIQUE INDEX "segments_slug_pSlugNull_gpSlugNull_tenantId_idx" ON "segments" ("slug", "tenantId") WHERE "grandparentSlug" IS NULL AND "parentSlug" IS NULL;

```
that guarantees slug uniqueness in project group and project levels

I'll try it on prod to see if there are any issues :+1: